Remove specific processing of declarations of add, sub, mul, div, rsub

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29238 Remove specific processing of declarations of add, sub, mul, div, rsub**
* #29602 Add functions add/sub/rsub/mul/div that accept a scalar as first arg

So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

xuhdev committed:
Remove specific processing of declarations of add, sub, mul, and div

So that their Scalar overload can be in effect from Python code.

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, and div"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

ezyang reviewed (approved):
assuming this passes CI seems fine

ailzhang reviewed (commented):
https://github.com/pytorch/pytorch/blob/cd9f8ab820d3ff15314ee070a38d8aad1412d4f6/torch/csrc/utils/python_arg_parser.cpp#L631-L655
This code blocks seem to be handling number->tensor conversion as well, maybe also worth removing?

xuhdev commented:
@ailzhang I'm not really sure about it. The function is used in https://github.com/pytorch/pytorch/blob/cd9f8ab820d3ff15314ee070a38d8aad1412d4f6/torch/csrc/utils/python_arg_parser.h#L224 , which looks like a legitimate function (e.g., `x = torch.tensor(3, dtype=torch.int)`). Maybe someone can shed some light on this?

xuhdev commented:
@ailzhang any idea about the XLA failure?

```
Nov 05 22:11:16 ======================================================================
Nov 05 22:11:16 ERROR: test_booleans (__main__.TestTypePromotion)
Nov 05 22:11:16 ----------------------------------------------------------------------
Nov 05 22:11:16 Traceback (most recent call last):
Nov 05 22:11:16   File "/var/lib/jenkins/workspace/xla/test/test_operations.py", line 1471, in test_booleans
Nov 05 22:11:16     self.assertEqual(torch.add(True, True), True)
Nov 05 22:11:16 TypeError: add() received an invalid combination of arguments - got (bool, bool), but expected one of:
Nov 05 22:11:16  * (Tensor input, Number other, Number alpha)
Nov 05 22:11:16  * (Tensor input, Tensor other, Number alpha, Tensor out)
Nov 05 22:11:16 
Nov 05 22:11:16 
Nov 05 22:11:16 ======================================================================
Nov 05 22:11:16 ERROR: test_comparison_ops_with_type_promotion (__main__.TestTypePromotion)
Nov 05 22:11:16 ----------------------------------------------------------------------
Nov 05 22:11:16 Traceback (most recent call last):
Nov 05 22:11:16   File "/var/lib/jenkins/workspace/xla/test/test_operations.py", line 1617, in test_comparison_ops_with_type_promotion
Nov 05 22:11:16     out_res = op['out_op'](t1, t2, device)
Nov 05 22:11:16   File "/var/lib/jenkins/workspace/xla/test/test_operations.py", line 1544, in <lambda>
Nov 05 22:11:16     x, y, out=torch.empty(1, dtype=torch.bool, device=d)),
Nov 05 22:11:16 RuntimeError: value cannot be converted to type uint8_t without overflow: 1.04858e+06
Nov 05 22:11:16 
Nov 05 22:11:16 ======================================================================
Nov 05 22:11:16 ERROR: test_from_issue (__main__.TestTypePromotion)
Nov 05 22:11:16 ----------------------------------------------------------------------
Nov 05 22:11:16 Traceback (most recent call last):
Nov 05 22:11:16   File "/var/lib/jenkins/workspace/xla/test/test_operations.py", line 1386, in test_from_issue
Nov 05 22:11:16     self.assertEqual((u + 1000).dtype, torch.uint8)  # integer overflow
Nov 05 22:11:16 RuntimeError: value cannot be converted to type uint8_t without overflow: 1000
Nov 05 22:11:16 
Nov 05 22:11:16 ======================================================================
Nov 05 22:11:16 FAIL: test_result_type (__main__.TestTypePromotion)
Nov 05 22:11:16 ----------------------------------------------------------------------
Nov 05 22:11:16 Traceback (most recent call last):
Nov 05 22:11:16   File "/var/lib/jenkins/workspace/xla/test/test_operations.py", line 1507, in test_result_type
Nov 05 22:11:16     torch.result_type(torch.tensor(1, dtype=torch.int), 1), torch.int)
Nov 05 22:11:16   File "/var/lib/jenkins/workspace/xla/test/test_operations.py", line 378, in assertEqual
Nov 05 22:11:16     super(XlaTestCase, self).assertEqual(x, y, message)
Nov 05 22:11:16 AssertionError: torch.int64 != torch.int32 : 
```

ailzhang reviewed (approved):
Yea this PR calls into a few more ops in XLA land, I'll get a patch in xla ready by tmr and coordinate the lands

gchanan commented:
Can you post the differences in generated code?  You can use tools/git_add_generated_dirs.

xuhdev commented:
@gchanan Here you go!

```diff
diff --git a/torch/csrc/autograd/generated/python_torch_functions.cpp b/torch/csrc/autograd/generated/python_torch_functions.cpp
index ed68256cfaa0..178425466903 100644
--- a/torch/csrc/autograd/generated/python_torch_functions.cpp
+++ b/torch/csrc/autograd/generated/python_torch_functions.cpp
@@ -1999,6 +1999,7 @@ static PyObject * THPVariable_add(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "add(Tensor input, Scalar other, Scalar alpha=1)",
     "add(Tensor input, Scalar alpha, Tensor other, *, Tensor out=None)|deprecated",
     "add(Tensor input, Tensor other, *, Scalar alpha=1, Tensor out=None)",
   }, /*traceable=*/true);
@@ -2007,12 +2008,14 @@ static PyObject * THPVariable_add(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_add(r.tensor(0), r.scalar(1), r.scalar(2)));
+  } else if (r.idx == 1) {
     if (r.isNone(3)) {
       return wrap(dispatch_add(r.tensor(0), r.scalar(1), r.tensor(2)));
     } else {
       return wrap(dispatch_add(r.tensor(0), r.scalar(1), r.tensor(2), r.tensor(3)));
     }
-  } else if (r.idx == 1) {
+  } else if (r.idx == 2) {
     if (r.isNone(3)) {
       return wrap(dispatch_add(r.tensor(0), r.tensor(1), r.scalar(2)));
     } else {
@@ -3958,6 +3961,7 @@ static PyObject * THPVariable_div(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "div(Tensor input, Scalar other)",
     "div(Tensor input, Tensor other, *, Tensor out=None)",
   }, /*traceable=*/true);
 
@@ -3965,6 +3969,8 @@ static PyObject * THPVariable_div(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_div(r.tensor(0), r.scalar(1)));
+  } else if (r.idx == 1) {
     if (r.isNone(2)) {
       return wrap(dispatch_div(r.tensor(0), r.tensor(1)));
     } else {
@@ -7191,6 +7197,7 @@ static PyObject * THPVariable_mul(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "mul(Tensor input, Scalar other)",
     "mul(Tensor input, Tensor other, *, Tensor out=None)",
   }, /*traceable=*/true);
 
@@ -7198,6 +7205,8 @@ static PyObject * THPVariable_mul(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_mul(r.tensor(0), r.scalar(1)));
+  } else if (r.idx == 1) {
     if (r.isNone(2)) {
       return wrap(dispatch_mul(r.tensor(0), r.tensor(1)));
     } else {
@@ -9565,6 +9574,7 @@ static PyObject * THPVariable_sub(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "sub(Tensor input, Scalar other, Scalar alpha=1)",
     "sub(Tensor input, Scalar alpha, Tensor other, *, Tensor out=None)|deprecated",
     "sub(Tensor input, Tensor other, *, Scalar alpha=1, Tensor out=None)",
   }, /*traceable=*/true);
@@ -9573,12 +9583,14 @@ static PyObject * THPVariable_sub(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_sub(r.tensor(0), r.scalar(1), r.scalar(2)));
+  } else if (r.idx == 1) {
     if (r.isNone(3)) {
       return wrap(dispatch_sub(r.tensor(0), r.scalar(1), r.tensor(2)));
     } else {
       return wrap(dispatch_sub(r.tensor(0), r.scalar(1), r.tensor(2), r.tensor(3)));
     }
-  } else if (r.idx == 1) {
+  } else if (r.idx == 2) {
     if (r.isNone(3)) {
       return wrap(dispatch_sub(r.tensor(0), r.tensor(1), r.scalar(2)));
     } else {
diff --git a/torch/csrc/autograd/generated/python_torch_functions_dispatch.h b/torch/csrc/autograd/generated/python_torch_functions_dispatch.h
index 8dc3d5fb5e44..be7cb7c22d62 100644
--- a/torch/csrc/autograd/generated/python_torch_functions_dispatch.h
+++ b/torch/csrc/autograd/generated/python_torch_functions_dispatch.h
@@ -548,6 +548,11 @@ inline std::tuple<Tensor,Tensor> dispatch_adaptive_max_pool1d(const Tensor & sel
   AutoNoGIL no_gil;
   return at::adaptive_max_pool1d(self, output_size);
 }
+inline Tensor dispatch_add(const Tensor & self, Scalar other, Scalar alpha) {
+
+  AutoNoGIL no_gil;
+  return self.add(other, alpha);
+}
 inline Tensor dispatch_add(const Tensor & self, Scalar alpha, const Tensor & other, Tensor out) {
 
   AutoNoGIL no_gil;
@@ -1453,6 +1458,11 @@ inline Tensor dispatch_dist(const Tensor & self, const Tensor & other, Scalar p)
   AutoNoGIL no_gil;
   return self.dist(other, p);
 }
+inline Tensor dispatch_div(const Tensor & self, Scalar other) {
+
+  AutoNoGIL no_gil;
+  return self.div(other);
+}
 inline Tensor dispatch_div(const Tensor & self, const Tensor & other, Tensor out) {
 
   AutoNoGIL no_gil;
@@ -2733,6 +2743,11 @@ inline std::tuple<Tensor,Tensor> dispatch_mode(const Tensor & self, int64_t dim,
   AutoNoGIL no_gil;
   return self.mode(dim, keepdim);
 }
+inline Tensor dispatch_mul(const Tensor & self, Scalar other) {
+
+  AutoNoGIL no_gil;
+  return self.mul(other);
+}
 inline Tensor dispatch_mul(const Tensor & self, const Tensor & other, Tensor out) {
 
   AutoNoGIL no_gil;
@@ -3773,6 +3788,11 @@ inline Tensor dispatch_stft(const Tensor & self, int64_t n_fft, c10::optional<in
   AutoNoGIL no_gil;
   return self.stft(n_fft, hop_length, win_length, window, normalized, onesided);
 }
+inline Tensor dispatch_sub(const Tensor & self, Scalar other, Scalar alpha) {
+
+  AutoNoGIL no_gil;
+  return self.sub(other, alpha);
+}
 inline Tensor dispatch_sub(const Tensor & self, Scalar alpha, const Tensor & other, Tensor out) {
 
   AutoNoGIL no_gil;
diff --git a/torch/csrc/autograd/generated/python_variable_methods.cpp b/torch/csrc/autograd/generated/python_variable_methods.cpp
index de6c822db723..9a014fb36720 100644
--- a/torch/csrc/autograd/generated/python_variable_methods.cpp
+++ b/torch/csrc/autograd/generated/python_variable_methods.cpp
@@ -1088,6 +1088,7 @@ static PyObject * THPVariable_add(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "add(Scalar other, Scalar alpha=1)",
     "add(Scalar alpha, Tensor other)|deprecated",
     "add(Tensor other, *, Scalar alpha=1)",
   }, /*traceable=*/true);
@@ -1096,8 +1097,10 @@ static PyObject * THPVariable_add(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
-    return wrap(dispatch_add(self, r.scalar(0), r.tensor(1)));
+    return wrap(dispatch_add(self, r.scalar(0), r.scalar(1)));
   } else if (r.idx == 1) {
+    return wrap(dispatch_add(self, r.scalar(0), r.tensor(1)));
+  } else if (r.idx == 2) {
     return wrap(dispatch_add(self, r.tensor(0), r.scalar(1)));
   }
   Py_RETURN_NONE;
@@ -1107,6 +1110,7 @@ static PyObject * THPVariable_add_(PyObject* self_, PyObject* args, PyObject* kw
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "add_(Scalar other, Scalar alpha=1)",
     "add_(Scalar alpha, Tensor other)|deprecated",
     "add_(Tensor other, *, Scalar alpha=1)",
   }, /*traceable=*/true);
@@ -1115,8 +1119,10 @@ static PyObject * THPVariable_add_(PyObject* self_, PyObject* args, PyObject* kw
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
-    return wrap(dispatch_add_(self, r.scalar(0), r.tensor(1)));
+    return wrap(dispatch_add_(self, r.scalar(0), r.scalar(1)));
   } else if (r.idx == 1) {
+    return wrap(dispatch_add_(self, r.scalar(0), r.tensor(1)));
+  } else if (r.idx == 2) {
     return wrap(dispatch_add_(self, r.tensor(0), r.scalar(1)));
   }
   Py_RETURN_NONE;
@@ -2221,6 +2227,7 @@ static PyObject * THPVariable_div(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "div(Scalar other)",
     "div(Tensor other)",
   }, /*traceable=*/true);
   auto& self = reinterpret_cast<THPVariable*>(self_)->cdata;
@@ -2228,6 +2235,8 @@ static PyObject * THPVariable_div(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_div(self, r.scalar(0)));
+  } else if (r.idx == 1) {
     return wrap(dispatch_div(self, r.tensor(0)));
   }
   Py_RETURN_NONE;
@@ -2237,6 +2246,7 @@ static PyObject * THPVariable_div_(PyObject* self_, PyObject* args, PyObject* kw
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "div_(Scalar other)",
     "div_(Tensor other)",
   }, /*traceable=*/true);
   auto& self = reinterpret_cast<THPVariable*>(self_)->cdata;
@@ -2244,6 +2254,8 @@ static PyObject * THPVariable_div_(PyObject* self_, PyObject* args, PyObject* kw
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_div_(self, r.scalar(0)));
+  } else if (r.idx == 1) {
     return wrap(dispatch_div_(self, r.tensor(0)));
   }
   Py_RETURN_NONE;
@@ -3914,6 +3926,7 @@ static PyObject * THPVariable_mul(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "mul(Scalar other)",
     "mul(Tensor other)",
   }, /*traceable=*/true);
   auto& self = reinterpret_cast<THPVariable*>(self_)->cdata;
@@ -3921,6 +3934,8 @@ static PyObject * THPVariable_mul(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_mul(self, r.scalar(0)));
+  } else if (r.idx == 1) {
     return wrap(dispatch_mul(self, r.tensor(0)));
   }
   Py_RETURN_NONE;
@@ -3930,6 +3945,7 @@ static PyObject * THPVariable_mul_(PyObject* self_, PyObject* args, PyObject* kw
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "mul_(Scalar other)",
     "mul_(Tensor other)",
   }, /*traceable=*/true);
   auto& self = reinterpret_cast<THPVariable*>(self_)->cdata;
@@ -3937,6 +3953,8 @@ static PyObject * THPVariable_mul_(PyObject* self_, PyObject* args, PyObject* kw
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
+    return wrap(dispatch_mul_(self, r.scalar(0)));
+  } else if (r.idx == 1) {
     return wrap(dispatch_mul_(self, r.tensor(0)));
   }
   Py_RETURN_NONE;
@@ -5381,6 +5399,7 @@ static PyObject * THPVariable_sub(PyObject* self_, PyObject* args, PyObject* kwa
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "sub(Scalar other, Scalar alpha=1)",
     "sub(Scalar alpha, Tensor other)|deprecated",
     "sub(Tensor other, *, Scalar alpha=1)",
   }, /*traceable=*/true);
@@ -5389,8 +5408,10 @@ static PyObject * THPVariable_sub(PyObject* self_, PyObject* args, PyObject* kwa
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
-    return wrap(dispatch_sub(self, r.scalar(0), r.tensor(1)));
+    return wrap(dispatch_sub(self, r.scalar(0), r.scalar(1)));
   } else if (r.idx == 1) {
+    return wrap(dispatch_sub(self, r.scalar(0), r.tensor(1)));
+  } else if (r.idx == 2) {
     return wrap(dispatch_sub(self, r.tensor(0), r.scalar(1)));
   }
   Py_RETURN_NONE;
@@ -5400,6 +5421,7 @@ static PyObject * THPVariable_sub_(PyObject* self_, PyObject* args, PyObject* kw
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
+    "sub_(Scalar other, Scalar alpha=1)",
     "sub_(Scalar alpha, Tensor other)|deprecated",
     "sub_(Tensor other, *, Scalar alpha=1)",
   }, /*traceable=*/true);
@@ -5408,8 +5430,10 @@ static PyObject * THPVariable_sub_(PyObject* self_, PyObject* args, PyObject* kw
   auto r = parser.parse(args, kwargs, parsed_args);
 
   if (r.idx == 0) {
-    return wrap(dispatch_sub_(self, r.scalar(0), r.tensor(1)));
+    return wrap(dispatch_sub_(self, r.scalar(0), r.scalar(1)));
   } else if (r.idx == 1) {
+    return wrap(dispatch_sub_(self, r.scalar(0), r.tensor(1)));
+  } else if (r.idx == 2) {
     return wrap(dispatch_sub_(self, r.tensor(0), r.scalar(1)));
   }
   Py_RETURN_NONE;
diff --git a/torch/csrc/autograd/generated/python_variable_methods_dispatch.h b/torch/csrc/autograd/generated/python_variable_methods_dispatch.h
index fb05763dce4b..beb7a2445d5c 100644
--- a/torch/csrc/autograd/generated/python_variable_methods_dispatch.h
+++ b/torch/csrc/autograd/generated/python_variable_methods_dispatch.h
@@ -169,6 +169,11 @@ inline Tensor dispatch_acos_(Tensor & self) {
   AutoNoGIL no_gil;
   return self.acos_();
 }
+inline Tensor dispatch_add(Tensor & self, Scalar other, Scalar alpha) {
+
+  AutoNoGIL no_gil;
+  return self.add(other, alpha);
+}
 inline Tensor dispatch_add(Tensor & self, Scalar alpha, const Tensor & other) {
 
   AutoNoGIL no_gil;
@@ -179,6 +184,11 @@ inline Tensor dispatch_add(Tensor & self, const Tensor & other, Scalar alpha) {
   AutoNoGIL no_gil;
   return self.add(other, alpha);
 }
+inline Tensor dispatch_add_(Tensor & self, Scalar other, Scalar alpha) {
+
+  AutoNoGIL no_gil;
+  return self.add_(other, alpha);
+}
 inline Tensor dispatch_add_(Tensor & self, Scalar alpha, const Tensor & other) {
 
   AutoNoGIL no_gil;
@@ -724,11 +734,21 @@ inline Tensor dispatch_dist(Tensor & self, const Tensor & other, Scalar p) {
   AutoNoGIL no_gil;
   return self.dist(other, p);
 }
+inline Tensor dispatch_div(Tensor & self, Scalar other) {
+
+  AutoNoGIL no_gil;
+  return self.div(other);
+}
 inline Tensor dispatch_div(Tensor & self, const Tensor & other) {
 
   AutoNoGIL no_gil;
   return self.div(other);
 }
+inline Tensor dispatch_div_(Tensor & self, Scalar other) {
+
+  AutoNoGIL no_gil;
+  return self.div_(other);
+}
 inline Tensor dispatch_div_(Tensor & self, const Tensor & other) {
 
   AutoNoGIL no_gil;
@@ -1479,11 +1499,21 @@ inline std::tuple<Tensor,Tensor> dispatch_mode(Tensor & self, int64_t dim, bool
   AutoNoGIL no_gil;
   return self.mode(dim, keepdim);
 }
+inline Tensor dispatch_mul(Tensor & self, Scalar other) {
+
+  AutoNoGIL no_gil;
+  return self.mul(other);
+}
 inline Tensor dispatch_mul(Tensor & self, const Tensor & other) {
 
   AutoNoGIL no_gil;
   return self.mul(other);
 }
+inline Tensor dispatch_mul_(Tensor & self, Scalar other) {
+
+  AutoNoGIL no_gil;
+  return self.mul_(other);
+}
 inline Tensor dispatch_mul_(Tensor & self, const Tensor & other) {
 
   AutoNoGIL no_gil;
@@ -2124,6 +2154,11 @@ inline Tensor dispatch_stft(Tensor & self, int64_t n_fft, c10::optional<int64_t>
   AutoNoGIL no_gil;
   return self.stft(n_fft, hop_length, win_length, window, normalized, onesided);
 }
+inline Tensor dispatch_sub(Tensor & self, Scalar other, Scalar alpha) {
+
+  AutoNoGIL no_gil;
+  return self.sub(other, alpha);
+}
 inline Tensor dispatch_sub(Tensor & self, Scalar alpha, const Tensor & other) {
 
   AutoNoGIL no_gil;
@@ -2134,6 +2169,11 @@ inline Tensor dispatch_sub(Tensor & self, const Tensor & other, Scalar alpha) {
   AutoNoGIL no_gil;
   return self.sub(other, alpha);
 }
+inline Tensor dispatch_sub_(Tensor & self, Scalar other, Scalar alpha) {
+
+  AutoNoGIL no_gil;
+  return self.sub_(other, alpha);
+}
 inline Tensor dispatch_sub_(Tensor & self, Scalar alpha, const Tensor & other) {
 
   AutoNoGIL no_gil;
```

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

gchanan reviewed (commented):
Can we fix this first?
```
>>> torch.add(1,1, out=2.)
tensor(2., dtype=torch.float64)
```

Or maybe yours fixes this?  Can you check?

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev commented:
@gchanan I think this PR has resolved this issue:

```
In [2]: torch.add(1,1, out=2.)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-4fd005df26c5> in <module>
----> 1 torch.add(1,1, out=2.)

TypeError: add() received an invalid combination of arguments - got (int, int, out=float), but expected one of:
 * (Tensor input, Tensor other, Number alpha, Tensor out)
 * (Number other, Tensor input, Number alpha)
      didn't match because some of the keywords were incorrect: out
 * (Tensor input, Number other, Number alpha)
      didn't match because some of the keywords were incorrect: out
```

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f79b6a99:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f79b6a99a06bf7f5833351d4eea7d7f98d91c21d).

### :detective: 3 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4285047) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (1/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87434842) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87434842))
<details>
<summary>
<code>Jan 16 23:29:16 /pytorch/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected ‘}’ at end of input</code>
</summary>

```
Jan 16 23:29:16 /pytorch/aten/src/ATen/native/BinaryOps.cpp:562:75: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:16  Tensor& logical_xor_out(Tensor& result, const Tensor& self, Scalar other) { return comparison_op_out(result, self, other, static_cast<OutFunc>(at::logical_xor_out)); } 
Jan 16 23:29:16                                                                            ^ 
Jan 16 23:29:16 /pytorch/aten/src/ATen/native/BinaryOps.cpp:563:54: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:16  Tensor logical_xor(const Tensor& self, Scalar other) { return comparison_op(self, other, static_cast<OutFunc>(at::logical_xor_out)); } 
Jan 16 23:29:16                                                       ^ 
Jan 16 23:29:16 /pytorch/aten/src/ATen/native/BinaryOps.cpp:564:50: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:16  Tensor& logical_xor_(Tensor& self, Scalar other) { return comparison_op_(self, other, static_cast<OutFunc>(at::logical_xor_out)); } 
Jan 16 23:29:16                                                   ^ 
Jan 16 23:29:16 /pytorch/aten/src/ATen/native/BinaryOps.cpp: At global scope: 
Jan 16 23:29:16 /pytorch/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected ‘}’ at end of input 
Jan 16 23:29:16  }  // namespace at 
Jan 16 23:29:16  ^ 
Jan 16 23:29:16 In file included from /pytorch/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 16 23:29:16                  from /pytorch/aten/src/ATen/Tensor.h:11, 
Jan 16 23:29:16                  from /pytorch/aten/src/ATen/Context.h:4, 
Jan 16 23:29:16                  from /pytorch/aten/src/ATen/ATen.h:5, 
Jan 16 23:29:16                  from /pytorch/aten/src/ATen/native/Col2Im.cpp:1: 
Jan 16 23:29:16 /pytorch/aten/src/ATen/native/Col2Im.cpp: In function ‘at::Tensor at::native::col2im_cpu(const at::Tensor&, c10::IntArrayRef, c10::IntArrayRef, c10::IntArrayRef, c10::IntArrayRef, c10::IntArrayRef)’: 
Jan 16 23:29:16 /pytorch/c10/core/MemoryFormat.h:32:46: warning: ‘c10::MemoryFormat c10::get_contiguous_memory_format()’ is deprecated [-Wdeprecated-declarations] 
Jan 16 23:29:16  #define LEGACY_CONTIGUOUS_MEMORY_FORMAT c10::get_contiguous_memory_format() 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4285053) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (2/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87434855) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87434855))
<details>
<summary>
<code>Jan 16 23:29:22 /pytorch/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected ‘}’ at end of input</code>
</summary>

```
Jan 16 23:29:22 /pytorch/aten/src/ATen/native/BinaryOps.cpp:562:75: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:22  Tensor& logical_xor_out(Tensor& result, const Tensor& self, Scalar other) { return comparison_op_out(result, self, other, static_cast<OutFunc>(at::logical_xor_out)); } 
Jan 16 23:29:22                                                                            ^ 
Jan 16 23:29:22 /pytorch/aten/src/ATen/native/BinaryOps.cpp:563:54: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:22  Tensor logical_xor(const Tensor& self, Scalar other) { return comparison_op(self, other, static_cast<OutFunc>(at::logical_xor_out)); } 
Jan 16 23:29:22                                                       ^ 
Jan 16 23:29:22 /pytorch/aten/src/ATen/native/BinaryOps.cpp:564:50: error: a function-definition is not allowed here before ‘{’ token 
Jan 16 23:29:22  Tensor& logical_xor_(Tensor& self, Scalar other) { return comparison_op_(self, other, static_cast<OutFunc>(at::logical_xor_out)); } 
Jan 16 23:29:22                                                   ^ 
Jan 16 23:29:22 /pytorch/aten/src/ATen/native/BinaryOps.cpp: At global scope: 
Jan 16 23:29:22 /pytorch/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected ‘}’ at end of input 
Jan 16 23:29:22  }  // namespace at 
Jan 16 23:29:22  ^ 
Jan 16 23:29:22 In file included from /pytorch/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 16 23:29:22                  from /pytorch/aten/src/ATen/Tensor.h:11, 
Jan 16 23:29:22                  from /pytorch/aten/src/ATen/Context.h:4, 
Jan 16 23:29:22                  from /pytorch/aten/src/ATen/ATen.h:5, 
Jan 16 23:29:22                  from /pytorch/aten/src/ATen/native/Col2Im.cpp:1: 
Jan 16 23:29:22 /pytorch/aten/src/ATen/native/Col2Im.cpp: In function ‘at::Tensor at::native::col2im_cpu(const at::Tensor&, c10::IntArrayRef, c10::IntArrayRef, c10::IntArrayRef, c10::IntArrayRef, c10::IntArrayRef)’: 
Jan 16 23:29:22 /pytorch/c10/core/MemoryFormat.h:32:46: warning: ‘c10::MemoryFormat c10::get_contiguous_memory_format()’ is deprecated [-Wdeprecated-declarations] 
Jan 16 23:29:22  #define LEGACY_CONTIGUOUS_MEMORY_FORMAT c10::get_contiguous_memory_format() 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4285049) pytorch_linux_xenial_py3_clang5_mobile_build (3/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=87435408) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=87435408))
<details>
<summary>
<code>Jan 16 23:30:11 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected '}' at end of input</code>
</summary>

```
Jan 16 23:30:11  Tensor& bitwise_xor_out(Tensor& result, const Tensor& self, const Tensor& other) { 
Jan 16 23:30:11                                                                                   ^ 
Jan 16 23:30:11 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:355:61: error: a function-definition is not allowed here before '{' token 
Jan 16 23:30:11  Tensor bitwise_xor(const Tensor& self, const Tensor& other) { 
Jan 16 23:30:11                                                              ^ 
Jan 16 23:30:11 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected '}' at end of input 
Jan 16 23:30:11  }  // namespace at 
Jan 16 23:30:11  ^ 
Jan 16 23:30:11 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp: At global scope: 
Jan 16 23:30:11 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected '}' at end of input 
Jan 16 23:30:11 /var/lib/jenkins/workspace/aten/src/ATen/native/BinaryOps.cpp:566:1: error: expected '}' at end of input 
Jan 16 23:30:11 caffe2/CMakeFiles/torch_cpu.dir/build.make:2296: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o' failed 
Jan 16 23:30:11 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/BinaryOps.cpp.o] Error 1 
Jan 16 23:30:11 make[2]: *** Waiting for unfinished jobs.... 
Jan 16 23:30:13 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Jan 16 23:30:13 CMakeFiles/Makefile2:903: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Jan 16 23:30:13 Makefile:138: recipe for target 'all' failed 
Jan 16 23:30:13 make: *** [all] Error 2 
Jan 16 23:30:13 =================== sccache compilation log =================== 
Jan 16 23:30:13 + sccache_epilogue 
Jan 16 23:30:13 + echo '=================== sccache compilation log ===================' 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

gchanan commented:
It looks like this was approved but never landed.

@xuhdev @ailzhang do we still want to land this?

xuhdev commented:
@gchanan Yes, I'm still actively fixing bugs on the stacked PR. Once the stacked PR is ready, this one should be ready to go

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev commented:
I have a question related to the current CI failure. On the master branch, I have

```python
In [1]: import torch

In [2]: torch.add(1, 2)
Out[2]: tensor(3)

In [3]: torch.tan(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-25eb9cba79db> in <module>
----> 1 torch.tan(1)

TypeError: tan(): argument 'input' (position 1) must be Tensor, not int
```

Is this inconsistency intended? Do we also want to illegalize `torch.add(1, 2)`-type of operations or change the behavior of `torch.tan(1)`?

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]

xuhdev committed:
Update on "Remove specific processing of declarations of add, sub, mul, div, rsub"


So that their Scalar overload can be in effect from Python code.

As discussed in https://github.com/pytorch/pytorch/pull/28534#issuecomment-546366009

[ghstack-poisoned]
